### ISSUE TITLE
Question About Loss Function Computation

### ISSUE THREAD

**Issue by Shellorley0513** [RAGEN-AI/RAGEN#62]
I found that in the update actor, RAGEN uses loss mask to mask the log_prob part related to observation. It seems to calculate the log_prob of only the tokens with loss_mask=1.
```
for data in micro_batches:
    data = data.cuda()  # actor device is cpu when using offload
    responses = data['responses']
    response_length = responses.size(1)
    attention_mask = data['attention_mask']
    response_mask = attention_mask[:, -response_length:]
    if self.config.state_masking:
        response_mask = data['loss_mask']
    old_log_prob = data['old_log_probs']
    advantages = data['advantages']

    clip_ratio = self.config.clip_ratio
    entropy_coeff = self.config.entropy_coeff

    # all return: (bsz, response_length)
    entropy, log_prob = self._forward_micro_batch(micro_batch=data, temperature=temperature)

    pg_loss, pg_clipfrac, ppo_kl = core_algos.compute_policy_loss(old_log_prob=old_log_prob,
                                                                  log_prob=log_prob,
                                                                  advantages=advantages,
                                                                  eos_mask=response_mask,
                                                                  cliprange=clip_ratio)
    # compute entropy loss from entropy
    entropy_loss = verl_F.masked_mean(entropy, response_mask)

    # compute policy loss
    policy_loss = pg_loss - entropy_loss * entropy_coeff

    if self.config.use_kl_loss:
        ref_log_prob = data['ref_log_prob']
        # compute kl loss
        kld = core_algos.kl_penalty(logprob=log_prob,
                                    ref_logprob=ref_log_prob,
                                    kl_penalty=self.config.kl_loss_type)
        kl_loss = masked_mean(kld, response_mask)

        policy_loss = policy_loss + kl_loss * self.config.kl_loss_coef
        metrics['actor/kl_loss'] = kl_loss.detach().item()
        metrics['actor/kl_coef'] = self.config.kl_loss_coef

    loss = policy_loss / self.gradient_accumulation
    loss.backward()
```
However, in def _create_loss_mask(), loss_mask is set to 1 for observation tokens and 0 for the assistant's output. 
```
def _create_loss_mask(self, batch, metrics):
    """Create loss mask for state tokens."""
    response_length = batch.batch['responses'].shape[-1]
    response_mask = batch.batch['attention_mask'][:, -response_length:]
    
    # Initialize state mask
    state_mask = torch.zeros_like(response_mask)
    
    responses = [self.tokenizer.decode(resp, skip_special_tokens=False) for resp in batch.batch['responses']]

    for i, response in enumerate(responses):
        # Find all pairs of start and end marker positions
        start_marker = self.config.algorithm.state_masking.start_state_marker
        end_marker = self.config.algorithm.state_masking.end_state_marker   
        
        # Get all start and end positions
        start_positions = [m.start() for m in re.finditer(re.escape(start_marker), response)]
        end_positions = [m.start() + len(end_marker) for m in re.finditer(re.escape(end_marker), response)]
        
        
        # Convert character positions to token positions
        for start, end in zip(start_positions, end_positions):
            prefix_to_start = response[:start]
            state_section = response[start:end]
            
            start_tokens = self.tokenizer.encode(prefix_to_start, add_special_tokens=False)
            state_tokens = self.tokenizer.encode(state_section, add_special_tokens=False)
            
            start_token_pos = len(start_tokens)
            end_token_pos = start_token_pos + len(state_tokens)
            
            state_mask[i, start_token_pos:end_token_pos] = 1
    
    loss_mask = state_mask * response_mask
    batch.batch['loss_mask'] = loss_mask
    
    # Debug print
    print("\nRaw batch[0] (before masking):\n", self.tokenizer.decode(batch.batch['responses'][0]))
    response_ids = batch.batch['responses'][0]
    unmasked_ids = response_ids[loss_mask[0] == 0]
    print("\nUnmasked batch[0] (after masking):\n", self.tokenizer.decode(unmasked_ids))
    
    masked_ids = response_ids[loss_mask[0] == 1]
    print("\nMasked batch[0] (masked parts):\n", self.tokenizer.decode(masked_ids))
    
    metrics.update({
        'state_tokens/total': loss_mask.sum().item(),
        'state_tokens/coverage': (loss_mask.sum() / response_mask.sum()).item(),
    })
    
    return batch, metrics
```
Would this be a mistake in loss computation? If loss_mask is meant to represent the observation tokens, should it be inverted to correctly mask out the observation during loss computation?

**Comment by ZihanWang314** [comment]
@JamesKrW

**Comment by jedward225** [comment]
problem solved, a pr would be merged these days(actually the bug is located in `_create_loss_mask`. previous implementation mistook the meaning of `0` token and `1` token)
